### DISCUSSION TITLE
Isochronous transfer - how to make it continuous?

### DISCUSSION THREAD

**Discussion by simntonic** [libusb/libusb#1531] [Q&A]
Hi :) 

Firstly a huge thank you to all developpers who made the libusb possible and freely available - much appreciated! :) 

Now to my issue: I thought the whole concept of isochronous transfers is a constant **uninterrupted** stream of data.
If we agree with this, then I must be doing something wrong because I cannot do that with libusb. I can submit one "serie" of iso transfers, for example 16 transfers of 96 bytes - that's great. But when these finish there is a small stream interruption before the next "serie" is submitted, that's because the callback is only called when the transfer is fully finished.

Normally you would get a callback **before** the transfer is fully finished so you can prepare and queue the next tranfer so the data flow is uninterrupted. Is there a way to do this with libusb? If not - and please excuse my question - but what's the point of having isochronous transfers at all? You might as well use bulk ;) 

Thank you in advance
Simon

**Comment by simntonic** [comment]
PS: here is my basic test code that sends isochronous packets. I get 16 USB frames with 96 bytes each - perfect - then 2 frames without data (the time for the callback to be processed), then again 16 frames with 96 bytes each, and so on. What I'm trying is to constantly send 96 bytes every frame, without interruption.

```
#include <stdio.h>
#include <libusb-1.0/libusb.h>

#define EP_ISO_OUT       0x01

#define VENDOR_ID        0x1234U
#define PRODUCT_ID       0x7853U

#define USB_ITF_ID       0

static volatile int do_exit = 0;
static struct libusb_device_handle *devh = NULL;

#define	BYTES_PER_FRAME		96 // 96 bytes per frame
#define	NUM_FRAMES_IN_BUF	16 // 16 frames in buffer

static uint8_t buf[BYTES_PER_FRAME*NUM_FRAMES_IN_BUF];
static struct libusb_transfer *xfr;

int counter = 0;
static void LIBUSB_CALL cb_xfr(struct libusb_transfer *curXfr)
{
	if (curXfr->status != LIBUSB_TRANSFER_COMPLETED) {
		fprintf(stderr, "Error: transfer status %d\n", curXfr->status);
		libusb_free_transfer(curXfr);
		do_exit = 3;
		return;
	}

	counter++;
	printf("Transfer %d completed, length:%u bytes\n", counter, curXfr->length);

	int rc = libusb_submit_transfer(curXfr);
	if (rc < 0) {
		fprintf(stderr, "Error re-submitting iso transfer: %s\n", libusb_strerror(rc));
		do_exit = 4;
		return;
	}
}

void fillDummyBuffer(void) {
	int i, j;
	for (i = 0; i < NUM_FRAMES_IN_BUF; i++) {
		for (j = 0; j < BYTES_PER_FRAME; j++)
			buf[i*BYTES_PER_FRAME+j] = (uint8_t)(i*BYTES_PER_FRAME+j);
	}
}


int main(void)
{
	int rc;

	xfr = libusb_alloc_transfer(NUM_FRAMES_IN_BUF);
	if (!xfr) {
		fprintf(stderr, "Cannot allocate memory for transfer!\n");
		return 2;
	}

	rc = libusb_init(NULL);
	if (rc != LIBUSB_SUCCESS) {
		fprintf(stderr, "Error initializing libusb: %s\n", libusb_strerror(rc));
		return 1;
	}

	devh = libusb_open_device_with_vid_pid(NULL, VENDOR_ID, PRODUCT_ID);
	if (!devh) {
		fprintf(stderr, "Error finding USB device from Vendor/Product ID\n");
		goto out;
	}

	rc = libusb_claim_interface(devh, USB_ITF_ID);
	if (rc != LIBUSB_SUCCESS) {
		fprintf(stderr, "Error claiming interface: %s\n", libusb_strerror(rc));
		goto out;
	}

	fillDummyBuffer();

	libusb_fill_iso_transfer(xfr, devh, EP_ISO_OUT, buf, BYTES_PER_FRAME*NUM_FRAMES_IN_BUF, NUM_FRAMES_IN_BUF, cb_xfr, NULL, 0);
	libusb_set_iso_packet_lengths(xfr, BYTES_PER_FRAME);
	libusb_submit_transfer(xfr);
	printf("Transfer submitted\n");

	while (!do_exit) {
		rc = libusb_handle_events(NULL);
		if (rc != LIBUSB_SUCCESS)
			break;
	}

	libusb_release_interface(devh, USB_ITF_ID);
	
  out:
	if (devh)
		libusb_close(devh);
	libusb_exit(NULL);
	return rc;
}
```

**Comment by simntonic** [comment]
Ok I actually found the solution myself - it would be really great if there was an example and/or documentation about how to do this. There is only one example using libusb for isochronous.

If someone is having the same issue: the solution is to use 2 transfers, as they can be submitted both at the same time (the second one is basically queued). Then, when the first one finishes the callback is called, and it can be resubmitted before the second one finishes. And so on - so the data flow is now uninterrupted :)

**Reply by simntonic** [reply]
Ok it turns out this does not work 100%.

It works about 80% of the time, then when the OS gets busy doing something else (for example open a new window, etc.), the flow gets interrupted because the libusb_handle_events() is not called often enough.


I have tried to improve the libusb_handle_events() by calling it inside of a very precise timer callback, which happens every 3ms with +/- 50us (measured), even when the OS is busy with GUI or other tasks. That function should obviously be as fast as possible, so I am calling libusb_handle_events_timeout() with a timeout of zero.

In pseudo code it's something like this:

```
libusb_handle_events_timeout( zero timeout )
    -> calls the transfer complete callback for each transfer X which is finished:
             set transfer[X] complete = true

for (i = 0, i < AMNT_TRANSFERS, i++) {
    if (transfer[i] complete) {
        libusb_submit_transfer(transfer[i])
        set transfer[i] complete = false
    }
}
```

This works about 99% of the time BUT sometimes - when the OS is busy - the libusb_handle_events_timeout() does not call the callback. Because of this the transfer complete flag stays at false, even though I can see the hardware device actually received the isochronous packet so the transfer is in fact complete.
This happens maybe a few more times in a row, and then finally all transfer complete flags are set to true. If I only have 2 or 3 transfers, all it takes is for this to happen 2-3 times in a row and the data flow gets interrupted.

Am I doing something wrong? I think this is something to do with the poll() and those file descriptors but I do not know the inner intricacies of the libusb to understand what I'm missing. All I want libusb_handle_events_timeout() is simply to check if the transfer is finished so I can start it again in time.

Has anyone every used libusb to send constant uninterrupted isochronous packets? I would love to see an example of it :)

Thank you
Simon

**Reply by brmarkus** [reply]
Not under MS-WIN... but under Linux (as well as in embedded environments, e.g. for automotive and USB camera streams in high resolutions), increassing the process' priority, and/or pin to a specific CPU core. Not sure how to change priority under MS-Win, pin a process to a dedicated CPU-core under MS-Win...
But to be honest not using `libusb`...

**Reply by simntonic** [reply]
I see - what do you use then?

I mean I have the dedicated high priority thread already (as mentioned above), all I need is a simple way to submit an isochronous transfer and check if it is completed :) Any suggestions?

**Reply by brmarkus** [reply]
It was using proprietary interfaces using a proprietary framework - but finally accessing USBFS kernel interfaces.

**Reply by tormodvolden** [reply]
There is no reason to limit it to 2 transfers, try for instance 8 transfers.

If you are running libusb_handle_events() in a tight loop (the function is blocking so that is fine), I don't see why a timer is better, if you give the event handling thread as high priority as you would give the timer.

Note our documentation https://libusb.sourceforge.io/api-1.0/group__libusb__poll.html#ga4989086e3f0327f3886a4c474ec7c327 recommends to use libusb_handle_events_completed().

**Reply by simntonic** [reply]
Thanks for the reply @tormodvolden !


> There is no reason to limit it to 2 transfers, try for instance 8 transfers.

Yes it's a very good point - I've already done this as you can see in the pseudo-code above (with `AMNT_TRANSFERS`). I've tried pushing it to 20 or even 100, and actually sometimes the flow still gets interrupted. It's all a matter of priority, sometimes if the computer is busy on graphics for example the USB events do not get marked as done (it's not about calling `libusb_handle_events()`, it's more internally in libusb - I'm not exactly sure at what level that is, my guess is it's running in userspace and not kernel space, hence the delay). The other problem with a huge amount of transfer is the delay, as I need 100 transfers planned "in advance", which means the flow can only start delayed and not real time.


> If you are running libusb_handle_events() in a tight loop (the function is blocking so that is fine), I don't see why a timer is better, if you give the event handling thread as high priority as you would give the timer.

I'm sorry to say but I disagree with this - basically you say I should run a blocking function in a high priority thread... That's just bad practice - otherwise how will the rest of the operating system work? Either you should have blocking stuff with low priority (like a typical main loop), or you have high priority thread which are meant to run as quickly as possible and return, so it does not prevent other high priority stuff from running on time.
I'm trying to implement the second option, running`libusb_handle_events_timeout(zero)` in a precise high priority timer (in a way it does not matter if it's a timer or not - the point is it's a high priority thread that will always run "often enough"). That seems to work fine, but like I said the problem is sometimes the function does not consider events done even though the actual data has been sent already, and if it does that for too long it breaks the continuity of the transfers.

**Reply by tormodvolden** [reply]
It is blocking because it performs a blocking system call, waiting for something arriving from the kernel. If nothing is happening, the kernel is not scheduling that thread - because there is nothing to do. So it leaves the CPU free to other processes, whatever priority they may have.
